src/askui_runner/modules/core/infrastructure/files/askui.py

```diff
@@ -41,13 +57,158 @@ def __init__(self, base_url: str, headers: dict[str, str]):
+            ),
+            "delete_remote": lambda remote_file_path: (
+                logging.info(f"Dry: Delete remote file {remote_file_path}")
+                if dry
+                else self._delete_file(remote_file_path)
+            ),
+            "skip": lambda relative_path: logging.info(
+                f"Skip {relative_path} (no changes)"
+            ),
+        }
+
+        for relative_path in sorted(all_paths):
+            local_file: Union[LocalFileDto | None] = local_files.get(relative_path)
+            remote_file: Union[FileDto | None] = remote_files.get(relative_path)
+            remote_file_path = f"{remote_dir_path}/{relative_path}".replace("\\", "/")
+
+            # File exists in both locations
+            if local_file and remote_file:
+                local_mtime = local_file.last_modified
+                remote_mtime = (
+                    datetime.strptime(remote_file.last_modified, "%Y-%m-%dT%H:%M:%SZ")
+                    .replace(tzinfo=timezone.utc)
+                    .timestamp()
+                )
+
+                if source_of_truth == "local" and (
+                    local_mtime > remote_mtime or local_file.size != remote_file.size
+                ):
+                    action_mapper["upload"](local_file.path, remote_file_path)
+
+                elif source_of_truth == "remote" and (
+                    remote_mtime > local_mtime or local_file.size != remote_file.size
+                ):
+                    local_path = os.path.join(local_dir_path, relative_path)
+                    action_mapper["download"](
+                        remote_file.url, remote_file.last_modified, local_path
+                    )
+                else:
+                    action_mapper["skip"](relative_path)
+                    continue
+
+            # File exists only locally
+            elif local_file:
+                if source_of_truth == "local":
+                    action_mapper["upload"](local_file.path, remote_file_path)
+                elif delete:
+                    action_mapper["delete_local"](local_file.path)
+
+            # File exists only remotely
+            elif remote_file:
+                if source_of_truth == "remote":
+                    local_path = os.path.join(local_dir_path, relative_path)
+                    action_mapper["download"](
+                        remote_file.url, remote_file.last_modified, local_path
+                    )
+                elif delete:
+                    action_mapper["delete_remote"](remote_file_path)
+
+    # region Private Functions
     @retry(stop=stop_after_attempt(5), wait=wait_exponential(), reraise=True)
     def _upload_file(self, local_file_path: str, remote_file_path: str) -> None:
         with open(local_file_path, "rb") as f:
             url = urljoin(
                 base=self.base_url + "/",
                 url=remote_file_path,
             )
+            url = quote(url, safe=":/")
```

Comment:
Why quote the url?  Wouldn't it make more sense to just quote the `remote_file_path` before passing it to `urljoin`?